Fix -Wthread-safety-analysis warnings. Compile with -Wthread-safety-analysis if available.

practicalswift opened the pull request:
* Add mutex requirement for `AddToCompactExtraTransactions(…)`.
* Use `-Wthread-safety-analysis` if available.
* Rebased on top of https://github.com/TheBlueMatt/bitcoin/commits/2017-08-test-10923 - now includes: Fix -Wthread-safety-analysis warnings. Change the sync.h primitives to std from boost.

TheBlueMatt commented:
Do we actually enforce that on Travis? Its nice to have the auto-checker stuff, but AFAIR its (barely) hooked up anywhere?

practicalswift commented:
Travis is currently not compiling with `-Wthread-safety-analysis`.

Personally I build daily with `-Wthread-safety-analysis` enabled.

We want to kill all warnings emitted when compiling with `-Wthread-safety-analysis`, don't we? :-)

sipa commented:
If we're going to revive this safety checking infrastructure (it was introduced in #2002, and I don't think updated much since), there are probably many other places to add annotations.

jonasschnelli commented:
Agree with @sipa.
@practicalswift whats the reason of only add the safety checking infrastructure to `AddToCompactExtraTransactions`?

practicalswift commented:
The reason I'm starting with `AddToCompactExtraTransactions(…)` is simply to fix the only currently warning when building with `clang`'s `-Wthread-safety-analysis` enabled:

```
net_processing.cpp:587:10: warning: reading variable 'vExtraTxnForCompact' requires holding mutex 'cs_main' [-Wthread-safety-analysis]
```

The reason for the warning is that `vExtraTxnForCompact` is marked as guarded by the `cs_main`-mutex using the `GUARDED_BY(cs_main)` annotation:

```c++
static std::vector<std::pair<uint256, CTransactionRef>> vExtraTxnForCompact 
    GUARDED_BY(cs_main);
```

`clang` detects that `vExtraTxnForCompact` is accessed in `AddToCompactExtraTransactions(…)` and that this function lacks the corresponding `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` annotation.

This PR fixes that and after applying this PR the code base is free from `-Wthread-safety-analysis` warnings.

I'd be glad to extend the usage of `GUARDED_BY(…)` annotations (and the corresponding `EXCLUSIVE_LOCKS_REQUIRED(…)` annotations) in the code base, but that is a larger undertaking that probably belongs in a follow-up PR.

I guess one reason not to fix the current warning is if we for some reason would want to remove this thread-safety infrastructure. Let me know if that is the case :-)

practicalswift commented:
Additional reading: The ["C/C++ Thread Safety Analysis"](https://research.google.com/pubs/archive/42958.pdf) paper (Hutchins, Ballman & Sutherland, 2014) paper describing clang thread-safety analysis and how it is used in the Google C++ codebase.

TheBlueMatt commented:
Ah! Didn't realize its to fix a warning, yea, lets a) do this and b) can you add a -Werror=thread-safety-analysis here, too? I was informed travis already does a clang build for OSX, so it should be able to catch errors here :).

practicalswift commented:
@TheBlueMatt Sure! I've updated the title to make things more clear :-)

TheBlueMatt commented:
I think we really should make travis enforce this. It shouldnt be hard cause we already have --enable-werror by default, just need to hook up -Werror=thread-safety-analysis if the compiler supports it (which our clang-osx build should).

practicalswift commented:
@TheBlueMatt Would something along the lines of the following do the trick?

```diff
diff --git a/.travis.yml b/.travis.yml
index a79428f..423b04e 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -34,7 +34,7 @@ env:
 # No wallet
     - HOST=x86_64-unknown-linux-gnu PACKAGES="python3" DEP_OPTS="NO_WALLET=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-glibc-back-compat --enable-reduce-exports"
 # Cross-Mac
-    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports" OSX_SDK=10.11 GOAL="deploy"
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports" OSX_SDK=10.11 GOAL="deploy" CPPFLAGS="-Werror=thread-safety-analysis"

 before_install:
     - export PATH=$(echo $PATH | tr ':' "\n" | sed '/\/opt\/python/d' | tr "\n" ":" | sed "s|::|:|g")
```

TheBlueMatt commented:
@practicalswift I was referring to the -Werror support we already have in configure.ac. Best to use that, I'd think.

practicalswift commented:
@TheBlueMatt Got it! What about simply using `-Wthread-safety-analysis` if provided by the compiler? Like this:

```diff
diff --git a/configure.ac b/configure.ac
index aea5d71..a48aaac 100644
--- a/configure.ac
+++ b/configure.ac
@@ -249,6 +249,7 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wformat],[CXXFLAGS="$CXXFLAGS -Wformat"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wvla],[CXXFLAGS="$CXXFLAGS -Wvla"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wformat-security],[CXXFLAGS="$CXXFLAGS -Wformat-security"],,[[$CXXFLAG_WERROR]])
+  AX_CHECK_COMPILE_FLAG([-Wthread-safety-analysis],[CXXFLAGS="$CXXFLAGS -Wthread-safety-analysis"],,[[$CXXFLAG_WERROR]])

   ## Some compilers (gcc) ignore unknown -Wno-* options, but warn about all
   ## unknown options if any other warning is produced. Test the -Wfoo case, and
```

Looks good? :-)

practicalswift committed:
Use -Wthread-safety-analysis if available (+ -Werror=thread-safety-analysis if --enable-werror)

TheBlueMatt commented:
Heh, I figured you'd incldue it in this PR. Anyway, utACK this pr by itself.

theuni commented:
utACK ef2aca6c12efc48ce557a87c90ecf2d96ad4885e

TheBlueMatt commented:
utACK ef2aca6c12efc48ce557a87c90ecf2d96ad4885e, though its useful to hold off merging this until #10923 is ready as a test for #10923.

TheBlueMatt committed:
Fix -Wthread-safety-analysis warnings. Change the sync.h primitives to std from boost.

Commit 1.

This code was written by @TheBlueMatt in the following branch:
* https://github.com/TheBlueMatt/bitcoin/commits/2017-08-test-10923

This commit message was written by me (@practicalswift) who also squashed
@TheBlueMatt's commits into one and tried to summarize the changes made.

Commit 2.

Remove boost include. Remove boost mentions in comments.

practicalswift commented:
Cherry-picked 9df85380aae4219920ccca80efac2bdf408519d8 into this PR as suggested by @TheBlueMatt in #10923!

TheBlueMatt commented:
@ptracticalswift maybe rebase on #10923 and take the patches from https://github.com/TheBlueMatt/bitcoin/commits/2017-08-test-10923 (squashed, of course)?

practicalswift commented:
@TheBlueMatt Does this PR look good now? Please check my attempt att summarizing your changes in: https://github.com/bitcoin/bitcoin/pull/10866/commits/695fe315e8c4191bd5e9e98a8ce6d2be8c6aff96. Let me know if anything should be changed or if you'd rather squash your commits in your branch and have me cherry pick the resulting commit as-is :-)

MarcoFalke commented on .travis.yml:
```diff
@@ -34,7 +34,7 @@ env:
     - HOST=x86_64-unknown-linux-gnu PACKAGES="python3" DEP_OPTS="NO_WALLET=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-glibc-back-compat --enable-reduce-exports"
 # Cross-Mac
-    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports" OSX_SDK=10.11 GOAL="deploy"
+    - HOST=x86_64-apple-darwin11 PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev" BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror" OSX_SDK=10.11 GOAL="deploy"
```
This change is already in master

practicalswift commented:
@MarcoFalke Thanks! Rebased on to of `master`, so the already merged commit is now excluded from this PR :-)

danra commented:
@practicalswift 
NACK
1) Transaction time check was changed:
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
 +            checktxtime = std::chrono::steady_clock::now() + std::chrono::seconds(1);

2) IMO there are too many things going on in the first commit at once, making it hard to review and bug-prone. Switching the primitives in the sync module from boost to std should be done in an atomic commit, without doing any other change.
2.1) I didn't know about this pull request before opening issue #11166 (from which you referred me here), I'm pretty close to having that ready to open a pull request, (just changing boost to std). Will update here when it's ready. I think it'd be better if that were reviewed and merged, and only then follow up with other changes.

practicalswift commented:
@danra Are you referring to @TheBlueMatt:s commit f01ea9326bc69d4f731a5904881740308c5e4271?

This PR contains three commits. Please consider reviewing d77d9ccd1ad35f550c8dcdc3b83e65bad8c51c48 and 0c7115c56fdbe25257aee23ae002dc5af3238bb7 too :-)

practicalswift commented:
@sipa @TheBlueMatt @jonasschnelli @theuni I have some follow up work which depends on this PR. Do you have time reviewing this PR?

The follow-up work is a PR adding `GUARDED_BY(…)` and `EXCLUSIVE_LOCKS_REQUIRED(…)` annotations where appropriate.

danra commented:
@practicalswift Yes, referring to f01ea93

practicalswift commented:
@TheBlueMatt Do you have any input on @danra:s review?

practicalswift commented:
Please see #11226 for the follow-up PR which adds `GUARDED_BY(…)` and `EXCLUSIVE_LOCKS_REQUIRED(…)` annotations.

TheBlueMatt commented:
@danra Indeed, the minutes -> seconds change was a mistake, should probably be fixed. Indeed, the first commit was originally several different commits, but they werent logically split up so it was easier to squash them than to manually re-split them. Personally I find the first commit not-too-hard to review cause the changes are pretty clearly along file boundaries, but then I also wrote it, so I cant speak....

utACK 0c7115c56fdbe25257aee23ae002dc5af3238bb7 (I went back and re-checked that nothing currently needing boost interruption support should be effected by this change).

practicalswift commented:
@TheBlueMatt I've now added a commit fixing the minutes/seconds mix-up. Looks good? :-)

TheBlueMatt commented:
Yup, needs a squash though.

practicalswift commented:
Anyone willing to review? :-)

Having this PR merged would simplify the work on follow-up PR:s such as #11226.

TheBlueMatt commented:
@practicalswift generally, when squashing, please try to avoid rebasing as it makes it much easier for reviewers to track which changes you made between two commits of a PR.

re-utACK b22210337da62047290fcb2658b2c45117c7efa9

theuni commented on src/rpc/mining.cpp:
```diff
@@ -472,17 +472,17 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
+                if (cvBlockChange.wait_until(lock, checktxtime) == std::cv_status::timeout)
```
cvBlockChange needs to be notified in Interrupt() in init, otherwise we'll hang here at shutdown.

practicalswift replied:
@theuni Something like this? :-)

```diff
diff --git a/src/init.cpp b/src/init.cpp
index 25b8caf..73f4860 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -167,6 +167,7 @@ void Interrupt(boost::thread_group& threadGroup)
     if (g_connman)
         g_connman->Interrupt();
     threadGroup.interrupt_all();
+    cvBlockChange.notify_all();
 }

 void Shutdown()
```

theuni replied:
Looking deeper, I think this is actually ok (though circuitous) as-is.
```
StopRPC() -> g_rpcSignals.Stopped() -> OnRPCStopped() -> cvBlockChange.notify_all()
```
So, nevermind my comment.

danra commented:
`Personally I find the first commit not-too-hard to review cause the changes are pretty clearly along file boundaries, but then I also wrote it, so I cant speak....`

IMHO the fact that there were a few utACKs before fixing the minutes-seconds bug demonstrates that the commit was doing too many things at once to review easily.

So it's still a NACK from me until it's split to smaller commits. Of course, you're welcome to use my own PR #11199 which does just the move from `boost` to `std` locks part.

theuni commented:
utACK other than my comment above. I think the commits here are ok. Either way, with #11199 or this one, we're going to need some follow-ups to tidy up afterwards.

practicalswift commented:
@TheBlueMatt Thanks for the utACK. I'll avoid rebasing when squashing going forward. Thanks for the suggestion!

@theuni Thanks for the utACK.

Anyone else who has time to review? Having this merged would help my annotations work a lot :-)

ryanofsky commented on src/sync.h:
```diff
@@ -97,21 +100,23 @@ class CCriticalSection : public AnnotatedMixin<boost::recursive_mutex>
 #endif
 
 /** Wrapper around boost::unique_lock<Mutex> */
-template <typename Mutex>
```
Comment above still mentions boost.

ryanofsky commented on src/sync.h:
```diff
@@ -97,21 +100,23 @@ class CCriticalSection : public AnnotatedMixin<boost::recursive_mutex>
 {
 private:
-    boost::unique_lock<Mutex> lock;
+    std::unique_lock<CCriticalSection> lock;
```
In commit "Fix -Wthread-safety-analysis warnings."

Comment above still mentions boost.

ryanofsky commented on src/sync.h:
```diff
@@ -97,21 +100,23 @@ class CCriticalSection : public AnnotatedMixin<boost::recursive_mutex>
 /** Wrapper around boost::unique_lock<Mutex> */
-template <typename Mutex>
-class SCOPED_LOCKABLE CMutexLock
+class SCOPED_LOCKABLE CCriticalBlock
```
In commit "Fix -Wthread-safety-analysis warnings."

The naming here doesn't seem very consistent:

Recursive mutex: **CCriticalSection** (unchanged)
Recursive lock: **CCriticalBlock** (previously _CMutexLock_)
Non-recursive mutex: **CWaitableCriticalSection** (unchanged)
Non-recursive lock: **CWaitableLock** (previously didn't exist)

I don't see why we wouldn't just follow the standard naming and our current coding convention and use:

Recursive mutex: **RecursiveMutex**
Recursive lock: **RecursiveLock**
Non-recursive mutex: **Mutex**
Non-recursive lock: **Lock**

Maybe defining backwards compatible _CCriticalSection_ and _CWaitableCriticalSection_ typedefs to avoid increasing the size of the diff.

ryanofsky reviewed (commented):
utACK b22210337da62047290fcb2658b2c45117c7efa9. I agree with others in not seeing a pressing need to split up the first commit since it is only ~20 lines, even though it would be a little nicer split up.

practicalswift commented:
@ryanofsky Thanks for the review. I've now added two commits addressing the two issues raised. Additionally I found a boost include that is no longer needed. Would you mind re-reviewing and confirming that your feedback was properly addressed? :-)

ryanofsky reviewed (commented):
utACK 7ab7662dc067951fc9fd1364bd6d1446c8770305

Thanks for cleanup. You could consider combining "Change the sync.h primitives" and "Use consistent naming" commits because the combined diff would be smaller, and having two renames sets of in one PR is a little confusing.

practicalswift commented:
@ryanofsky Good point regarding combining the related commits. Could you please re-review? :-)

theuni commented:
utACK 7ab7662dc067951fc9fd1364bd6d1446c8770305. I suspect that we may run into symbol collisions at some point by using the generic "Mutex" name, but it'll be obvious if that happens.

ryanofsky reviewed (commented):
utACK 0a5c7596567d3d1d0296cf831057c79454f7851d (just squashed 7ab7662dc067951fc9fd1364bd6d1446c8770305)

ryanofsky commented on src/init.cpp:
```diff
@@ -545,14 +545,14 @@ static void BlockNotifyCallback(bool initialSync, const CBlockIndex *pBlockIndex
 
 static bool fHaveGenesis = false;
-static boost::mutex cs_GenesisWait;
+static CWaitableCriticalSection cs_GenesisWait;
```
Could s/CWaitableCriticalSection/Mutex

TheBlueMatt replied:
Hmm, that feels weird? We now have CCriticalSection and Mutex which mean two different things, but its not clear /what/ they mean without looking at what they're typedef'd to? Can we just do this as a scripted diff in a separate commit?

ryanofsky replied:
> Hmm, that feels weird? We now have CCriticalSection and Mutex which mean two different things, but its not clear /what/ they mean without looking at what they're typedef'd to? Can we just do this as a scripted diff in a separate commit?

Scripted-diff renaming followup is at: https://github.com/bitcoin/bitcoin/pull/11599

ryanofsky commented on src/sync.h:
```diff
@@ -85,33 +87,35 @@ void static inline DeleteLock(void* cs) {}
-/** Wrapper around boost::unique_lock<Mutex> */
-template <typename Mutex>
-class SCOPED_LOCKABLE CMutexLock
+/** Wrapper around std::unique_lock<Mutex> */
```
Comment is a little misleading since it's a wrapper around std::unique_lock\<RecursiveMutex>

ryanofsky reviewed (commented):
Two minor comments (feel free to ignore)

practicalswift commented:
@ryanofsky Thanks for re-reviewing! Feedback address. Would you mind re-reviewing again? :-)

ryanofsky reviewed (commented):
It looks like you accidentally squashed the two changes suggested above into the second commit ("Add mutex requirement...") instead of the first commit.

Anyway, utACK 0f5341f644796ff88234fa1179012f1cedae5a76. Only difference is the two changes.

practicalswift committed:
Add mutex requirement for AddToCompactExtraTransactions(…)

The vector `vExtraTxnForCompact`, which is guarded by the mutex
`cs_main`, is accessed in `AddToCompactExtraTransactions(…)`.

practicalswift commented:
@ryanofsky Sorry, fixed. Would you mind re-reviewing? :-)

ryanofsky reviewed (commented):
utACK 94568c8f22e5969fd3abfca1c2afa7b80a05cf9a. No changes other than rearranged commits

TheBlueMatt commented:
Not sure how I feel about renaming CMutexLock type in the same commit that changes a bunch of stuff from boost -> std, let alone in a commit people were complaining was already too large.

practicalswift commented:
@TheBlueMatt Should I do a `s/CMutexLock/RecursiveLock/` or should I keep the rename but have it in a separate commit? Please advice.

ryanofsky commented:
CMutexLock is only mentioned in 5 lines of code, all of them in sync.h. Sometimes it's nice to make changes in separate commits, but I think this particular rename makes more sense in context, and doesn't seem make the PR more complicated to review. But maybe I'm missing something because this entire change is only 47 lines and seems pretty straightforward to me.

practicalswift commented:
@TheBlueMatt @ryanofsky I'll let you guys find consensus. I'll happily adjust :-)

TheBlueMatt commented:
@ryanofsky in this case the PR adds a compatibility typedef to allow places to keep using CCriticalSection, renaming the underlying type and then mapping the old typename to the new one. I agree it should be renamed, but we should actually rename it instead of halfway-renaming it (using scripted-diff =D).

ryanofsky commented:
Agree scripted diff would improve upon the typedef.

TheBlueMatt commented:
In the interest of getting this very-important PR merged sooner rather than later, can we skip it for now?

ryanofsky commented:
SGTM

theuni commented:
+1

practicalswift commented:
Please review the latest commit added - does that revert the rename as suggested?

ryanofsky commented:
@practicalswift, I think you can drop the "Revert CMutexLock rename" commit.

What Matt is asking you to do is remove the two typedefs:

```diff
+typedef RecursiveMutex CCriticalSection;
+typedef Mutex CWaitableCriticalSection
```

Which will require replacing RecursiveMutex to CCriticalSection and Mutex to CWaitableCriticalSection. You can squash these changes into the "Fix -Wthread-safety-analysis warnings" commit, which should make it smaller.

practicalswift commented:
@ryanofsky Could you take a look at `Revert [wip]`? :-)

ryanofsky commented:
That looks close. Here is another commit that builds on top of yours and gets rid of more unnecessary renames & diffs: 17ebad7b47a2567a75e5a701780a261bf1a3275d (I was wrong above when I asked you to drop CMutexLock commit). If you squash both of these into "Fix -Wthread-safety-analysis warnings" and edit the commit message to remove mention of renames, this should be good to go.

ryanofsky commented:
Here is a combined branch: https://github.com/ryanofsky/bitcoin/commits/pr/locks

You can fetch & apply it with

```sh
git fetch -n https://github.com/ryanofsky/bitcoin pr/locks:ryanofsky-locks
git reset --hard ryanofsky-locks
```

practicalswift commented:
@ryanofsky Looks good now? :-)

ryanofsky reviewed (commented):
Looks good! utACK 1864bea78ae306e91dde46fcc2b85a01999efd84

TheBlueMatt commented on src/sync.h:
```diff
@@ -85,33 +87,35 @@ void static inline DeleteLock(void* cs) {}
+typedef std::condition_variable CConditionVariable;
+
+/** Just a typedef for std::unique_lock, can be wrapped later if desired */
+typedef std::unique_lock<std::mutex> Lock;
```
I'd really rather we *not* call this a "Lock" - its not intended to be the thing used across the codebase as any lock (though we could maybe do so if we add DEBUG_LOCKORDER support here).

ryanofsky replied:
You always say what you don't want instead of saying what you do want ;). What would you like instead? If you've seen https://github.com/bitcoin/bitcoin/pull/11599, the name should make more sense in this context. I think Lock and Mutex are the right names because they match the standard. As you mentioned, you can implement DEBUG_LOCKORDER with them and also do other things useful for debugging, like printing differences in a data structure between time a particular mutex is locked & unlocked.

TheBlueMatt replied:
The original name (CWaitableCriticalSection) was nice and specific - I'm happy to make it more in-line with the more-accepted "lock" terminology, but pointing out that it has only a narrow use-case in our codebase is important.

ryanofsky replied:
Ok, but I don't think non-recursive locks should have a narrow usecase in our codebase. I thought everybody agreed non-recursivelocks should be preferred to recursive locks.

ryanofsky replied:
Also CWaitableCriticalSection was not the original name for this. There was no original name for this. Previously code just used boost::lock directly.

TheBlueMatt replied:
Heh, I meant in the original patch in this PR. I mean we use locks in lots of different ways, for new subsystems that have very careful locking models, obviously non-recursive locks are preferred, but I'm skeptical every new thing we add will use them.

practicalswift replied:
@ryanofsky @TheBlueMatt Could you provide a diff with your suggested change?

ryanofsky replied:
I think Matt basically wants the typedef to be called ScaryLock instead of Lock. But I'm fine with whatever is proposed here given there is already a PR open for renaming.

practicalswift replied:
OK, I'll fix. @TheBlueMatt, what is your suggested name instead of `Lock`? :-)

TheBlueMatt replied:
I was fine with the WaitableCriticalSection (or WaitableLock or whatever) version that was there originally.

practicalswift replied:
@TheBlueMatt @ryanofsky The following diff would work?

```diff
diff --git a/src/init.cpp b/src/init.cpp
index dd64602..9e3eb6b 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -551,7 +551,7 @@ static void BlockNotifyGenesisWait(bool, const CBlockIndex *pBlockIndex)
 {
     if (pBlockIndex != nullptr) {
         {
-            Lock lock_GenesisWait(cs_GenesisWait);
+            WaitableLock lock_GenesisWait(cs_GenesisWait);
             fHaveGenesis = true;
         }
         condvar_GenesisWait.notify_all();
@@ -1634,7 +1634,7 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)

     // Wait for genesis block to be processed
     {
-        Lock lock(cs_GenesisWait);
+        WaitableLock lock(cs_GenesisWait);
         while (!fHaveGenesis) {
             condvar_GenesisWait.wait(lock);
         }
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index 8fae01b..0ba0e96 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -478,7 +478,7 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
         {
             checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);

-            Lock lock(csBestBlock);
+            WaitableLock lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
                 if (cvBlockChange.wait_until(lock, checktxtime) == std::cv_status::timeout)
diff --git a/src/sync.h b/src/sync.h
index 66b6d96..20556af 100644
--- a/src/sync.h
+++ b/src/sync.h
@@ -105,7 +105,7 @@ typedef AnnotatedMixin<std::mutex> CWaitableCriticalSection;
 typedef std::condition_variable CConditionVariable;

 /** Just a typedef for std::unique_lock, can be wrapped later if desired */
-typedef std::unique_lock<std::mutex> Lock;
+typedef std::unique_lock<std::mutex> WaitableLock;

 #ifdef DEBUG_LOCKCONTENTION
 void PrintLockContention(const char* pszName, const char* pszFile, int nLine);
```

ryanofsky replied:
> @TheBlueMatt @ryanofsky The following diff would work?

Yes for me. (As I mentioned in my last comment I am happy with any naming in this pr since #11599 can clean up any inconsistencies.)

practicalswift replied:
@ryanofsky Excellent! OK with you @TheBlueMatt? :-)

TheBlueMatt replied:
Yea, looks good, thanks!

practicalswift commented:
@ryanofsky @TheBlueMatt Name change patch applied. Squashed and pushed. Could you please re-review? :-)

TheBlueMatt commented:
utACK 76ea17c7964c15dd90e10c2c257cdeb5847b3d69

ryanofsky reviewed (commented):
utACK 76ea17c7964c15dd90e10c2c257cdeb5847b3d69, confirmed only change since last review is name change.

sipa commented:
utACK 76ea17c7964c15dd90e10c2c257cdeb5847b3d69. Nice to see some further unboostification.
